### PR TITLE
Entities

### PR DESCRIPTION
**Pull request by WhiteBlackGoose**
Added entities for derivative, limit, integral. Merged with limits. Division improved. Simplificator improved

### CODE REVIEW

**Comment by Happypig375** [comment]
Hey! I'm writing reviews...

**Comment by WhiteBlackGoose** [comment]
You can still leave it I guess

File: AngouriMath/Functions/Utils/Utils.cs [hunk 1]
```diff
@@ -28,6 +28,11 @@ namespace AngouriMath.Functions
 {
     internal static class Utils
     {
+        private static readonly EDecimal LowerForGood = EDecimal.FromDecimal(1e-6m);
+        private static readonly EDecimal UpperForGood = EDecimal.FromDecimal(1e+8m);
+        internal static bool IsGoodAsDouble(EDecimal num)
+            => num.CompareTo(LowerForGood) > 0 && num.CompareTo(UpperForGood) < 0;
+
```
Comment (Happypig375) [inline AngouriMath/Functions/Utils/Utils.cs:35]: This will not work as noted in the comment about `System.Math` above.

File: README.md [hunk 2]
```diff
@@ -1,18 +1,18 @@
-[![Codacy Badge](https://api.codacy.com/project/badge/Grade/1e172cdf699645b59567032dd1ae5cab)](https://www.codacy.com/manual/Angourisoft/MathS?utm_source=github.com&amp;utm_medium=referral&amp;utm_content=Angourisoft/MathS&amp;utm_campaign=Badge_Grade)
```
Comment (Happypig375) [inline README.md:1]: Yeah, Codacy definitely is too restrictive over style choices.

File: README.md [hunk 3]
```diff
@@ -1,18 +1,18 @@
...
-![Test](https://github.com/asc-community/AngouriMath/workflows/Test/badge.svg)
-![Nuget](https://img.shields.io/nuget/dt/AngouriMath?color=blue&label=NuGet%20installs&logoColor=blue)
-![GitHub](https://img.shields.io/github/license/AngouriSoft/MathS?color=purple)
+## AngouriMath
 
+|                          |               |
```
Comment (Happypig375) [inline README.md:3]:
```suggestion
| Contributions welcome?   |  ALWAYS!      |
```

File: Tests/DotnetBenchmark/Algebra.cs [hunk 4]
```diff
@@ -27,6 +30,18 @@ public AlgebraTest()
...
+        private EDecimal dec = 3;
+        private EDecimal coef = EDecimal.FromDecimal(0.2m) ?? 0;
 
+        [Benchmark]
+        public void Cos()
+        {
+            dec = dec.Negate(MathS.Settings.DecimalPrecisionContext) ?? 0;
+            dec = dec.Multiply(dec, MathS.Settings.DecimalPrecisionContext) ?? 0;
+            dec = dec.Multiply(coef) ?? 0;
+            dec = dec.Cos(MathS.Settings.DecimalPrecisionContext) ?? 0;
```
Comment (Happypig375) [inline Tests/DotnetBenchmark/Algebra.cs:44]: Why is the null-coalescing operator here? If these operations ever return null, we should throw early.

File: Tests/UnitTests/Convenience/LatexTest.cs [hunk 5]
```diff
@@ -315,5 +315,23 @@ public class LatexTests
...
+        [TestMethod] public void Derivative2() =>
+            Test(@"\frac{d\left[x + y\right]}{d\left[quack\right]}", MathS.Derivative("x + y", "quack"));
+        [TestMethod] public void Derivative3() =>
+            Test(@"\frac{d^{3}\left[x + 1\right]}{dx^{3}}", MathS.Derivative("x + 1", "x", 3));
+        [TestMethod] public void Integral1() =>
+            Test(@"\int \left[x+1\right] dx", MathS.Integral("x + 1", "x"));
+        [TestMethod] public void Integral2() =>
+            Test(@"\int \int \left[x+1\right] dx dx", MathS.Integral("x + 1", "x", 2));
+        [TestMethod] public void Integral3() =>
+            Test(@"\int \left[x+1\right] d\left[xf\right]", MathS.Integral("x + 1", "xf"));
```
Comment (Happypig375) [inline Tests/UnitTests/Convenience/LatexTest.cs:329]: Should add a few more tests that test the `Stringize` -> `Latexize` fixes above.

File: Tests/UnitTests/PatternsTest/SimplifyTest.cs [hunk 6]
```diff
@@ -88,5 +89,18 @@ void AssertSimplify(Entity original, Entity simplified, int? level = null)
         [TestMethod] public void NaNP1() => AssertSimplify(nan + 1, nan);
         [TestMethod] public void NaN0() => AssertSimplify(nan * 0, nan);
         [TestMethod] public void NaNPow0() => AssertSimplify(MathS.Pow(nan, 0), nan);
+        [TestMethod] public void Derive1() => AssertSimplify(MathS.Derivative("x + 2", "x"), 1);
+        [TestMethod] public void Derive2() => AssertSimplify(MathS.Derivative("7x2 - x + 2", "x", 2), 14);
+        [TestMethod] public void Integral1() => AssertSimplify(MathS.Integral("x + y", "x", 0), "x + y");
+        [TestMethod] public void Divide1() => AssertSimplify("(x2 + 2 x y + y2) / (x + y)", "x + y");
+        [TestMethod] public void Divide2() => AssertSimplify("(x3 + 3 x 2 y + 3 x y 2 + y3) / (x + y)", "x2 + 2 x y + y2".Simplify());
+        [TestMethod] public void Divide3() => AssertSimplify("(x2 + 2 x y + y2 + 1) / (x + y)", "x + 1 / (x + y) + y".Simplify());
```
Comment (Happypig375) [inline Tests/UnitTests/PatternsTest/SimplifyTest.cs:97]: `Simplify` being called on both sides does not ensure that the simplified result is acceptable - we should maintain an expected result that does not change with the implementation and update it when needed. Test should be more restrictive to actually catch bugs.

File: README.md (hunk 2 above)
Comment (WhiteBlackGoose) [reply README.md:1]: Rather stupid and glitchy, to be honest.

File: Tests/DotnetBenchmark/Algebra.cs (hunk 4 above)
Comment (WhiteBlackGoose) [reply Tests/DotnetBenchmark/Algebra.cs:44]: They're never null, just didn't want to remove <WarningsAsErrors>

File: AngouriMath/Convenience/MathS.cs [hunk 7]
```diff
@@ -79,61 +90,95 @@ public static Set SolveEquation(Entity equation, VariableEntity var)
...
-        /// <param name="x"></param>
-        /// <returns></returns>
+        /// <param name="@base">
+        /// base-node of logarithm
+        /// </param>
+        /// <param name="x">
+        /// Argument-node of logarithm
+        /// </param>
+        /// <returns>
+        /// Logarithm-node
```
[the remark this answers is not in the document]
Comment (WhiteBlackGoose) [reply AngouriMath/Convenience/MathS.cs:132]: Iirc now they all have dashes, no? Anyway, too minor, I better keep working on something more major

File: AngouriMath/Convenience/MathS.cs [hunk 8]
```diff
@@ -143,176 +188,193 @@ public static Set SolveInequalityNumerically(Entity inequality, VariableEntity v
...
         /// <summary>
         /// <a href="https://en.wikipedia.org/wiki/Trigonometric_functions"/>
         /// </summary>
-        /// <param name="a"></param>
+        /// <param name="a">
+        /// Argument node of which Secant will be taken
+        /// </param>
         /// <returns>
-        /// C / B
+        /// Secant node
```
[the remark this answers is not in the document]
Comment (WhiteBlackGoose) [reply AngouriMath/Convenience/MathS.cs:219]: Oops 😄

File: README.md (hunk 3 above)
Comment (WhiteBlackGoose) [reply README.md:3]: -____-

Comment (Happypig375) [reply README.md:3]: 😏

File: Tests/DotnetBenchmark/Algebra.cs (hunk 4 above)
Comment (Happypig375) [reply Tests/DotnetBenchmark/Algebra.cs:44]: Nullable reference types is only enabled for AngouriMath the project. Or you want them to be enabled everywhere? Can be done with a `Directory.Build.props`.

File: AngouriMath/Convenience/MathS.cs (hunk 7 above)
[the remark this answers is not in the document]
Comment (Happypig375) [reply AngouriMath/Convenience/MathS.cs:132]: No. See below like "Secant node".

File: AngouriMath/Core/Sys/TrigonometricTableValues.cs [hunk 9]
```diff
@@ -144,7 +144,6 @@ private static EDecimal TwoPiOver(int a)
                 (TwoPiOver(6), f1_2 * Sqrt(3)),
                 (TwoPiOver(7), Sqrt(1 - MathS.Pow(f1_6 * (-1 + Cbrt((7 + 21 * Sqrt(-3)) / 2) + Cbrt((7 - 21 * Sqrt(-3)) / 2)), 2))),
                 (TwoPiOver(8), f1_2 * Sqrt(2)),
-                (TwoPiOver(9), i / 2 * (Cbrt((-1 - Sqrt(-3)) / 2) - Cbrt((-1 + Sqrt(-3)) / 2))),
```
[the remark this answers is not in the document]
Comment (Happypig375) [reply AngouriMath/Core/Sys/TrigonometricTableValues.cs:147]: @WhiteBlackGoose ?

File: AngouriMath/Functions/Output/Latex.cs [hunk 10]
```diff
@@ -229,6 +230,102 @@ internal static string Latex(List<Entity> args)
...
+        {
+            MathFunctions.AssertArgs(args.Count, 3);
+            var pow = args[2];
+            var powerIfNeeded = pow == IntegerNumber.One ? "" : "^{" + pow.Latexise() + "}";
+
+            var varOverDeriv = (args[1] is VariableEntity && args[1].Name.Length == 1
+                ? args[1].Name
+                : @"\left[" + args[1].Stringize(false) + @"\right]");
+
+            return @"\frac{d" + powerIfNeeded +
```
[the remark this answers is not in the document]
Comment (Happypig375) [reply AngouriMath/Functions/Output/Latex.cs:246]: @WhiteBlackGoose ?

File: AngouriMath/Functions/Output/Latex.cs [hunk 11]
```diff
@@ -229,6 +230,102 @@ internal static string Latex(List<Entity> args)
...
+                return args[0].Latexise(false);
+
+            var sb = new StringBuilder();
+            for (int i = 0; i < asInt; i++)
+                sb.Append(@"\int ");
+            sb.Append(@"\left[");
+            sb.Append(args[0].Latexise(false));
+            sb.Append(@"\right]");
+
+            // TODO: can we write d^2 x or (dx)^2 instead of dx dx?
```
[the remark this answers is not in the document]
Comment (Happypig375) [reply AngouriMath/Functions/Output/Latex.cs:279]: @WhiteBlackGoose ?

File: AngouriMath/Functions/Utils/PeterONumbersExtensions.cs [hunk 12]
```diff
@@ -70,6 +71,10 @@ public static EInteger Lcm(this EInteger bigintFirst, EInteger bigintSecond)
         public static EDecimal Cos(this EDecimal x, EContext context)
         {
             if (!x.IsFinite) return EDecimal.NaN;
+            
+            // TODO: this check should be here to improve the performance a little bit, but tests won't work with that
+            // if (Utils.IsGoodAsDouble(x)) return EDecimal.FromDouble(Math.Cos(x.ToDouble()));
```
[the remark this answers is not in the document]
Comment (Happypig375) [reply AngouriMath/Functions/Utils/PeterONumbersExtensions.cs:76]: @WhiteBlackGoose ?

File: Tests/UnitTests/PatternsTest/SimplifyTest.cs (hunk 6 above)
Comment (Happypig375) [reply Tests/UnitTests/PatternsTest/SimplifyTest.cs:97]: @WhiteBlackGoose ?

File: AngouriMath/Core/Sys/TrigonometricTableValues.cs (hunk 9 above)
[the remark this answers is not in the document]
Comment (Happypig375) [reply AngouriMath/Core/Sys/TrigonometricTableValues.cs:147]: @WhiteBlackGoose ??

File: AngouriMath/Functions/Output/Latex.cs (hunk 10 above)
[the remark this answers is not in the document]
Comment (Happypig375) [reply AngouriMath/Functions/Output/Latex.cs:246]: @WhiteBlackGoose ??

**Pull request merged**